test/functional/lib/redirect-handler/blockOrObserve.test.ts

```diff
@@ -30,9 +30,41 @@ const dynamicRulesConditions = (regexFilter) => ({
   ]
 })
 
+const TEST_TAB_ID = 1234
+const LAST_GROUP_REGEX = '((?:[^\\.]|$).*)$'
+
+/**
+ * Ensures that the tab is redirected to the given url on the first request.
+ *
+ * @param url
+ */
+function ensureTabRedirected (url): void {
+  expect(browserMock.tabs.query.called).to.be.true
+  expect(browserMock.tabs.update.called).to.be.true
+  expect(browserMock.tabs.update.firstCall.args).to.deep.equal([TEST_TAB_ID, { url }])
+}
+
+/**
+ * Ensures that the declarativeNetRequest API is called with the expected rule.
+ * @param expectedCondition
+ * @param regexSubstitution
+ */
+function ensureDeclrativeNetRequetRuleIsAdded ({ expectedCondition, regexSubstitution }): void {
+  expect(browserMock.declarativeNetRequest.updateDynamicRules.called).to.be.true
+  const [{ addRules, removeRuleIds }] = browserMock.declarativeNetRequest.updateDynamicRules.firstCall.args
+  expect(removeRuleIds).to.deep.equal([])
+  expect(addRules).to.have.lengthOf(1)
```

Comment:
is there ever a chance when adding a rule that a rule will also be removed?

Can we add these to the arguments for this helper function?

Also, can we move this helper function outside of this test file?